torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -341,8 +377,7 @@ def __init__(
         self._partition_parameters_cache: List[List[Dict]] = []
         self._index_to_param_cache: List[torch.Tensor] = []
         self._device_to_params_per_rank_cache: Dict[torch.device, List[List[torch.Tensor]]] = {}
-        self._device_to_buckets_cache: Dict[torch.device, List[List[_DDPBucket]]] = {}
```

Comment:
I originally used a mapping with `torch.device` as the key to emulate the non-overlapped case (`self._device_to_params_per_rank`). However, I realized that that kind of mapping is only required as a result of the normal parameter partitioning algorithm, where parameters assigned to a rank are not organized by device, and is not relevant when overlapping since DDP ensures that each gradient bucket only corresponds to parameters on the single device. 

This change is nice since now `self._buckets` does not have a different type depending on `overlap_with_ddp=True` or not and since `self._device_to_device_index` is not needed anymore.